src/server/engine_shard_set.cc

```diff
@@ -39,8 +49,260 @@ using absl::GetFlag;
+  return op_type & ObjectTypeForScan::ALL || (FromRedisType(redis_type) & op_type) != 0;
+}
+
+struct MemoryDefragParams {
+  mi_heap_t* heap_ptr = nullptr;
+  DbSlice* slice = nullptr;
+  uint64_t cursor = 0;
+  uint8_t object_type = 0u | ObjectTypeForScan::STRING;  // use with map for type
+
+  MemoryDefragParams(mi_heap_t* hp, DbSlice* s) : heap_ptr{hp}, slice{s} {
+  }
+
+  void AssignType(ObjectTypeForScan t) {
+    object_type |= static_cast<uint8_t>(t);
+  }
+};
+
 vector<EngineShardSet::CachedStats> cached_stats;  // initialized in EngineShardSet::Init
 
+enum class MemDefragState : uint8_t {
+  REQUIRED,
+  NOT_REQUIRED,
+};
+
+enum class MemUtilizationState : uint8_t { UNKOWN, UNDER, OVER };
+
+// few issues here:
+// 1. Do we need to lock? (I think we do, since this is a task that I'm not sure is running
+// exclusively). - No we don't need a lock
+// 2. If we do need to lock, we can use the code similar to scan - ScanCb function. This require the
+// DB index, which we don't have! - we would be using index 0
+// 3. We need to find a way to map between the memory that we are moving to the new location that we
+// want to copy to
+//    we have a function that can tell use about memory usage for the given pointer (which in this
+//    case can be either the key or the value!), so are we going to move the value? (I think so),
+//    the key? (not sure), both?
+//  No we don't need, when we are doing new allocation, it would ensure that the allocation will
+//  happen to the first available space, this would result in moving all the memory to the "front"
+//  of the list.
+// 4, This leave the biggest issue, we need to find a new location to move to, how? - No we don't
+// see above To summarized:
+//  IF I understand this correctly:
+//    for each database id on this shared: (this is only 0)
+//      lock each key - don't
+//      check memory utilization for the address  of the key/value -> only for the value
+//      move the key/value to new location  -> just do allocation (for the value), move data to new
+//      memory location and free old one.
+//   The for each is running by the task itself, and the cursor is the for index - again for now
+//   only index 0!
+
+auto ReallocateValueImpl(PrimeValue& current_value) -> bool {
+  // we know that this is a string, we checked that before
+  auto val = current_value.ToString();
+  // we can now reset the old object as we are going to create a new one
+  current_value.Reset();
+  // This should allocate a new value
+  current_value.SetString(val);
+  return current_value.MallocUsed() > 0;
+}
+
+auto ReallocateValue(PrimeIterator it, void* from, const MemoryDefragParams& params, DbIndex index)
+    -> bool {
+  // allocate a new memory for the value, and then free the memory that used for the old one
+  CHECK(it->second.ObjType() == OBJ_STRING);  // only supporting for string right now!
+  if (it->second.IsExternal()) {
```

Comment:
no need to support IsExternal can be part of the check above.

Reply:
this code should actually not be part of the production code, I added this just as part of me tests. I will move it